glean-core/src/ping/mod.rs

```diff
@@ -67,12 +70,80 @@ impl PingMaker {
         current_seq as usize
     }
 
+    fn get_start_end_times(&self, glean: &Glean, storage_name: &str) -> (Metric, Metric) {
+        let time_unit = TimeUnit::Minute;
+
+        let start_time = DatetimeMetric::new(
+            CommonMetricData {
+                name: format!("{}#start", storage_name),
+                category: "".into(),
+                send_in_pings: vec![INTERNAL_STORAGE.into()],
+                lifetime: Lifetime::User,
+                ..Default::default()
+            },
+            time_unit,
+        );
+
+        let end_time = DatetimeMetric::new(
+            CommonMetricData {
+                name: format!("{}#end", storage_name),
+                category: "".into(),
+                send_in_pings: vec![INTERNAL_STORAGE.into()],
+                lifetime: Lifetime::User,
+                ..Default::default()
+            },
+            time_unit,
+        );
+
+        let start_time_data = match StorageManager.snapshot_metric(
+            glean.storage(),
+            INTERNAL_STORAGE,
+            &start_time.meta().identifier(),
+        ) {
+            Some(Metric::Datetime(t, tu)) => {
+                log::info!("datetime start found: {:?}", t);
+                Metric::Datetime(t, tu)
+            }
+            _ => {
+                let now: DateTime<Local> = Local::now();
+                let naive = now.naive_utc();
+                let fixed_tz = Local.offset_from_utc_datetime(&naive);
+                let now_fixed: DateTime<FixedOffset> = fixed_tz.from_utc_datetime(&naive);
+                log::info!("datetime start missing, creating now: {:?}", now_fixed);
+                Metric::Datetime(now_fixed, time_unit)
```

Comment:
This should not really be `now`, but the time glean-core was initialized. See the [glean-ac implementation](https://github.com/mozilla-mobile/android-components/blob/e2a019f6f8ffd3d2fbf514af0f79fa7d2c21cc25/components/service/glean/src/main/java/mozilla/components/service/glean/ping/PingMaker.kt#L83-L92). I'm afraid we were lacking test coverage for that, so it would be great if you could add it here ;)